### DISCUSSION TITLE
Enable Child Supervision

### DISCUSSION THREAD

**Discussion by ksathara** [openthread/openthread#9900] [Q&A]
Hi,
our system is made out of one FTD as a coap server and many SEDs as coap clients. We have seen rarely some clients drop off from the child table of the FTD. The clients battery seems ok. Also the RSSI seems well within budget. 

Since we couldn't figure out what is causing this, as a secondary measure, we are planning to implement 'Child Supervision'.
I have few questions.

1. It is mentioned to enable 'OPENTHREAD_CONFIG_CHILD_SUPERVISION_ENABLE', in child_supervision.h in SDK. I dont think changing SDK content is a good idea as if the SDK is updated, we will loose all that. Is there a way to implement child supervision from application side?

2. What is to be done from FTD side?

3. What is to be done from SED side?

Thanks,
Kaushalya

**Comment by jwhui** [comment]
Child Supervision feature is now enabled by default since:
- https://github.com/openthread/openthread/pull/7993

**Comment by abtink** [comment]
Child Supervision was earlier a custom enhancement/feature provided by the OpenThread stack. It is now adopted by the Thread spec and is a required feature in the Thread 1.3.0 Consequently, in the `main` OT repo, you'll no longer find the `OPENTHREAD_CONFIG_CHILD_SUPERVISION_ENABLE` configuration option (see #7993).  If you update to later OpenThread stack commits, this feature will be enabled and used.

---

While slightly outdated, this guide may still offer insights: 
- [Child Supervision](https://github.com/openthread/ot-docs/blob/main/site/en/guides/build/features/child-supervision.md). 
- (the source code [`thread/child_supervision.hpp`](https://github.com/openthread/openthread/blob/main/src/core/thread/child_supervision.hpp))

---

> I dont think changing SDK content is a good idea as if the SDK is updated, we will loose all that. Is there a way to implement child supervision from application side?

Technically possible to do so, but I won't recommend it.

**Reply by ksathara** [reply]
Thanks. The issue is the SDK I use, which is Nordic SDK 2.3.0, it seems to be disabled by default. Maybe newer SDKs may be inline with Thread 1.3.0, but I am hesitant to change the SDK unless there is very good reason to do so.

I have enabled it via the OPENTHREAD_CONFIG_CHILD_SUPERVISION_ENABLE for the time, to check this function. I build both FTD and SED from scratch and I can see the sections in 'child_supervision.cpp' enabled.

My observations so far
1. I cant see 'Sending supervision message to child' on FTD console. The apparently send by the host when it sends the supervision message to the SED.
2. When I set supervision interval in FTD 5sec and supervision timeout in SED 10sec, I would hope there shouldn't be any supervision timeout in SED. But I see  'ChildSupervsn-: Supervision timeout. No frame from parent in 10 sec' on the SED console.

What is going on here?

Also is there a way to force a child out from FTD's child table to check the supervision behavior?

Thanks,
Kaushalya

**Reply by jwhui** [reply]
[Nordic SDK 2.3.0 documentation](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/2.3.0/nrf/protocols/thread/prebuilt_libs.html) seems to indicate that the Child Supervision feature is enabled.

OpenThread does not currently provide an API to evict a child table entry.

**Reply by ksathara** [reply]
Hi,
The SED doesn't seem to get the child supervision packets send by its paired FTD, when it has a router in the middle, who is 'not' supporting child supervision. I see my SED logs say it's child supervision times out. But when the SED is directly connected to the child supervision enabled FTD, everything works. Doesn't the parent's child supervision message routes through other routers with this feature disabled?
Cheers,
Kaushalya

**Reply by abtink** [reply]
Child supervision operates directly between a parent and its child, which always have a direct link to each other. 

If a parent node uses an older version without supervision support, the feature will not function. In this case, the child should automatically fall back to periodically sending "Child Update Requests" to maintain connectivity with its parent (the child supervision feature implements this fallback).

**Reply by ksathara** [reply]
Hi Abtink,
In a multi hop scenario, the parent means the router to which the SED is immediately connected? Or in other words the router who has the SED in it's child table?
Cheers,
Kaushalya

**Reply by jwhui** [reply]
The parent is the the router that the child directly attaches to.

https://openthread.io/guides/thread-primer/node-roles-and-types

**Comment by ksathara** [comment]
Hi guys,

Is this a child supervision message send by a parent? The parent is 0x2c00 (02:5b:eb:ab:b8:58:00:6e) and the child is 0x2c01 (d2:fc:2e:43:b4:c2:e8:76).
![image](https://github.com/openthread/openthread/assets/20408595/641b939e-d961-4137-8924-bad15c7d359c)

Cheers,
Kaushalya

**Reply by jwhui** [reply]
Yes, the parent sends Child Supervision messages to validate that the child still exists.
